src/cpp/cache.cpp

```diff
@@ -0,0 +1,349 @@
+            itemsList_.emplace_front(std::make_pair(key, value));
+            stat.first->second = itemsList_.begin();
+        }
+
+        if (capacity_ > 0 && itemsList_.size() > capacity_)
+            removeLast();
+    }
+
+    sol::optional<Value> get(const Key& key) {
+        const auto mapIter = itemsMap_.find(key);
+        if (mapIter != itemsMap_.end()) {
+            itemsList_.splice(itemsList_.begin(), itemsList_, mapIter->second);
+            return mapIter->second->second;
+        }
+        return sol::nullopt;
+    }
+
+    void clear() {
+        for (const auto& pair: itemsList_)
+            destructor_(pair.second);
+
+        itemsMap_.clear();
+        itemsList_.clear();
+    }
+
+    bool remove(const Key& key) {
+        const auto iter = itemsMap_.find(key);
+        if (iter == itemsMap_.end())
+                return false;
+
+        destructor_(iter->second->second);
+        itemsList_.erase(iter->second);
+        itemsMap_.erase(iter);
+        return true;
+    }
+
+    size_t size() {
+        return itemsList_.size();
+    }
+
+    void setCapacity(size_t capacity) {
+        capacity_ = capacity;
+
+        while (capacity > 0 && itemsList_.size() > capacity)
+            removeLast();
+    }
+
+    size_t capacity() {
+        return capacity_;
+    }
+
+private:
+    void removeLast() {
+        const auto& pair = itemsList_.back();
+        destructor_(pair.second);
+        itemsMap_.erase(pair.first);
+        itemsList_.pop_back();
+    }
+
+    typedef std::list<std::pair<Key, Value>> ItemsList;
+
+    size_t capacity_;
+    Destructor destructor_;
+    ItemsList itemsList_;
+    std::unordered_map<Key, typename ItemsList::iterator, Hash> itemsMap_;
+};
+
+struct HandleCacheContext {
+    Function object;
+    LuaRef   ref;
+};
+
+class LuaRegDestructor {
+public:
+    LuaRegDestructor(lua_State* lua) : lua_(lua) {}
+
+    void operator() (LuaRef ref) {
+        luaL_unref(lua_, LUA_REGISTRYINDEX, ref);
+    }
+
+    void operator() (const HandleCacheContext& ctx) {
+        luaL_unref(lua_, LUA_REGISTRYINDEX, ctx.ref);
+    }
+
+private:
+    lua_State* lua_;
+};
+
+struct FunctionHash {
+    size_t operator()(const Function& func) const {
+        return std::hash<GCHandle>{}(func.handle());
+    }
+};
+
+struct StateCachePair {
+    LruCache<Function, LuaRef, LuaRegDestructor, FunctionHash> handleCache;
+    LruCache<const void*, HandleCacheContext, LuaRegDestructor> functionCache;
+
+    StateCachePair(size_t capacity, const LuaRegDestructor& destructor)
+        : handleCache(destructor, capacity), functionCache(destructor, capacity)
+    {}
+};
+
+StateCachePair& getStateCache(lua_State* lua) {
+    assert(thisThreadCacheRef != LUA_NOREF);
+
+    sol::object obj(lua, sol::ref_index(thisThreadCacheRef));
+    return obj.as<StateCachePair>();
+}
+
+void removeStateCache(lua_State* lua) {
+    DEBUG("cache") << "remove Lua state " << std::hex << lua;
```

Comment:
Too may `DEBUG` statements will make code output unreadable.

Reply:
For that purposes we can debug to now =)